### PR TITLE
[jit] Add lazy script decorator

### PR DESCRIPTION
**Pull request by driazati**
Stacked PRs
 * #34938 - [jit] Remove stray `@script`
 * **#34935 - [jit] Add lazy script decorator**

Some users maintain libraries of code that is largely trace-able but not
script-able. However, some functions may need to be `@torch.jit.script`ed if
they contain control flow so the tracer will use the compiler version.
This however impacts library start up time as in #33418, so this PR adds
a workaround in the form of a `@torch.jit. _script_if_tracing `
that will only initialize the compiler if the function is called while
actually tracing.

Differential Revision: [D20569778](https://our.internmc.facebook.com/intern/diff/20569778/)

### CODE REVIEW

File: test/test_jit.py [hunk 1]
```diff
@@ -4641,6 +4641,35 @@ def forward(self, x):
...
+
+        FileCheck().check_not("goodbye").check_not("hello").run(traced_bad.graph)
+
+        # Working example
+        untraceable = torch.jit._lazy_script_while_tracing(untraceable)
+
+        def fn2(x):
+            return untraceable(x)
+
+        with self.capture_stdout():
```
Comment (eellison) [inline test/test_jit.py]: why capture stdout? you can @suppress_warnings if warnings are being supressed ?

File: test/test_jit.py [hunk 2]
```diff
@@ -4641,6 +4641,35 @@ def forward(self, x):
...
+            traced = torch.jit.trace(fn, [torch.ones(2, 2)])
+
+        FileCheck().check_not("goodbye").check_not("hello").run(traced.graph)
```
Comment (eellison) [inline test/test_jit.py]: shouldn't this be `check` and not `check_not` ?

File: torch/csrc/jit/python/init.cpp [hunk 3]
```diff
@@ -734,6 +734,7 @@ void initJITBindings(PyObject* module) {
       return op->schema();
     });
   });
+  m.def("_is_tracing", []() { return jit::tracer::isTracing(); });
```
Comment (eellison) [inline torch/csrc/jit/python/init.cpp]: we may already have an api for this, `torch.jit.is_tracing`, or something.

File: torch/jit/__init__.py [hunk 4]
```diff
@@ -1322,6 +1322,34 @@ def interface(obj):
...
+    def wrapper(*args):
+        if not torch._C._is_tracing():
+            # Not tracing, don't do anything
+            return fn(*args)
+
+        if 'fn' not in compiled_fn:
+            compiled_fn['fn'] = script(fn, _frames_up=1)
+        return compiled_fn['fn'](*args)
+
+    wrapper._torchscript_original_fn = fn
```
Comment (eellison) [inline torch/jit/__init__.py]: `wrapper._torchscript_original_fn = fn` isn't used

**Review by eellison — APPROVED** [review]
LGTM. One of your tests was wrong so if that's not working this may not work right now.

**Comment by eellison** [comment]
Also, not sure `_lazy_script` is the most clear. Maybe: `jit._script_if_tracing` ? `jit._compile_if_tracing` ?

**Comment by driazati** [comment]
`_script_if_tracing` sounds best, I'll change it

File: test/test_jit.py (hunk 1 above)
Comment (driazati) [reply test/test_jit.py]: there are prints in the test

File: test/test_jit.py (hunk 2 above)
Comment (driazati) [reply test/test_jit.py]: yeah, this was only passing since the inline mode was wrong

File: torch/csrc/jit/python/init.cpp (hunk 3 above)
Comment (driazati) [reply torch/csrc/jit/python/init.cpp]: I didn't see one in `dir(torch.jit)`

File: torch/jit/__init__.py [hunk 5]
```diff
@@ -1941,6 +1967,14 @@ def linear(x):
     """
     return False
 
+
+def is_tracing():
+    """
+    Returns ``True`` in tracing (if a function is called during the tracing of
+    code with ``torch.jit.trace``) and ``False`` otherwise.
+    """
+    return torch._C._is_tracing
```
Comment (mpariente) [inline torch/jit/__init__.py:1976]: Returns the function, and not the bool `torch._C._is_tracing()`. `_script_if_tracing` doesn't work on 1.6.0 but is fixed on master.

Comment (mpariente) [reply torch/jit/__init__.py:1976]: Fixed [here](https://github.com/pytorch/pytorch/pull/42486) for reference.

**Pull request closed without merging**